File: src/report.rs
//! Report rendering for different output formats.

use crate::types::{ConformanceMatrix, KernelReport, TestCategory, TestResult};

/// Render a report to terminal with colors.
pub fn render_terminal(report: &KernelReport) -> String {
    let mut output = String::new();

    // Header
    output.push_str(&format!(
        "\n{} Conformance Report: {} ({})\n",
        "=".repeat(60),
        report.kernel_name,
        report.implementation
    ));
    output.push_str(&format!(
        "Language: {} | Protocol: {} | Duration: {:?}\n",
        report.language, report.protocol_version, report.total_duration
    ));
    output.push_str(&format!("{}\n\n", "=".repeat(60)));

    // Results by tier
    for tier in [
        TestCategory::Tier1Basic,
        TestCategory::Tier2Interactive,
        TestCategory::Tier3RichOutput,
        TestCategory::Tier4Advanced,
    ] {
        let tier_results = report.tier_results(tier);
        if tier_results.is_empty() {
            continue;
        }

        let (passed, total) = report.tier_score(tier);
        output.push_str(&format!(
            "Tier {}: {} ({}/{})\n",
            tier.tier_number(),
            tier.description(),
            passed,
            total
        ));
        output.push_str(&format!("{}\n", "-".repeat(50)));

        for record in tier_results {
            let symbol = record.result.symbol();
            let emoji = record.result.emoji();
            output.push_str(&format!(
                "  {} {} {} ({:?})\n",
                emoji, symbol, record.name, record.duration
            ));

            // Show failure reason
            if let TestResult::Fail { reason, .. } = &record.result {
                output.push_str(&format!("      Reason: {}\n", reason));
            }
            if let TestResult::PartialPass { score, notes } = &record.result {
                output.push_str(&format!("      Score: {:.0}% - {}\n", score * 100.0, notes));
            }
        }
        output.push('\n');
    }

    // Summary
    output.push_str(&format!("{}\n", "=".repeat(60)));
    output.push_str(&format!(
        "Total: {}/{} ({:.0}%)\n",
        report.passed(),
        report.total(),
        report.score() * 100.0
    ));

    output
}

/// Render a report as JSON.
pub fn render_json(report: &KernelReport) -> String {
    serde_json::to_string_pretty(report).unwrap_or_else(|e| format!("{{\"error\": \"{}\"}}", e))
}

/// Render a matrix as JSON.
pub fn render_matrix_json(matrix: &ConformanceMatrix) -> String {
    serde_json::to_string_pretty(matrix).unwrap_or_else(|e| format!("{{\"error\": \"{}\"}}", e))
}

/// Render a single report as markdown.
pub fn render_markdown(report: &KernelReport) -> String {
    let mut output = String::new();

    output.push_str(&format!(
        "# {} Conformance Report\n\n",
        report.kernel_name
    ));
    output.push_str(&format!(
        "- **Implementation**: {}\n",
        report.implementation
    ));
    output.push_str(&format!("- **Language**: {}\n", report.language));
    output.push_str(&format!(
        "- **Protocol Version**: {}\n",
        report.protocol_version
    ));
    output.push_str(&format!(
        "- **Score**: {}/{} ({:.0}%)\n\n",
        report.passed(),
        report.total(),
        report.score() * 100.0
    ));

    // Results table
    output.push_str("| Test | Tier | Result | Duration |\n");
    output.push_str("|------|------|--------|----------|\n");

    for record in &report.results {
        let result_str = match &record.result {
            TestResult::Pass => "PASS".to_string(),
            TestResult::Fail { reason, .. } => format!("FAIL: {}", truncate(reason, 30)),
            TestResult::Unsupported => "SKIP".to_string(),
            TestResult::Timeout => "TIMEOUT".to_string(),
            TestResult::PartialPass { score, .. } => format!("PARTIAL ({:.0}%)", score * 100.0),
        };

        output.push_str(&format!(
            "| {} | {} | {} | {:?} |\n",
            record.name,
            record.category.tier_number(),
            result_str,
            record.duration
        ));
    }

    output
}

/// Render a matrix as a markdown comparison table.
pub fn render_matrix_markdown(matrix: &ConformanceMatrix) -> String {
    if matrix.reports.is_empty() {
        return "No reports in matrix.".to_string();
    }

    let mut output = String::new();

    output.push_str("# Kernel Conformance Matrix\n\n");
    output.push_str(&format!(
        "Generated: {}\n\n",
        matrix.generated_at.format("%Y-%m-%d %H:%M:%S UTC")
    ));

    // Get all test names
    let test_names = matrix.all_test_names();

    // Header row
    output.push_str("| Test |");
    for report in &matrix.reports {
        output.push_str(&format!(" {} |", report.kernel_name));
    }
    output.push('\n');

    // Separator row
    output.push_str("|------|");
    for _ in &matrix.reports {
        output.push_str("------|");
    }
    output.push('\n');

    // Data rows
    for test_name in test_names {
        output.push_str(&format!("| {} |", test_name));
        for report in &matrix.reports {
            let result = report
                .results
                .iter()
                .find(|r| r.name == test_name)
                .map(|r| r.result.emoji())
                .unwrap_or("-");
            output.push_str(&format!(" {} |", result));
        }
        output.push('\n');
    }

    // Summary row
    output.push_str("| **Score** |");
    for report in &matrix.reports {
        output.push_str(&format!(
            " {}/{} |",
            report.passed(),
            report.total()
        ));
    }
    output.push('\n');

    output
}

fn truncate(s: &str, max_len: usize) -> String {
    if s.len() <= max_len {
        s.to_string()
    } else {
        format!("{}...", &s[..max_len - 3])
    }
}

File: site/src/types/report.ts
/**
 * TypeScript types matching the Rust types in src/types.rs
 */

/** Classification of why a test failed, to help identify root cause */
export type FailureKind =
  | 'timeout'
  | 'protocol_error'
  | 'unexpected_message_type'
  | 'unexpected_content'
  | 'kernel_error'
  | 'harness_error';

/** Human-readable hints for failure kinds */
export const FAILURE_HINTS: Record<FailureKind, string> = {
  timeout: 'Kernel may be slow to start or not responding. Try increasing timeout.',
  protocol_error: 'Message parsing failed. Check runtimed protocol crate for compatibility.',
  unexpected_message_type: 'Kernel sent wrong message type. Check kernel implementation.',
  unexpected_content: 'Response format differs from spec. Check kernel implementation.',
  kernel_error: 'Kernel reported an error. Check kernel logs for details.',
  harness_error: 'Test harness issue. Check test setup and dependencies.',
};

/** Likely source of the failure */
export const FAILURE_SOURCES: Record<FailureKind, string> = {
  timeout: 'kernel',
  protocol_error: 'runtimed',
  unexpected_message_type: 'kernel',
  unexpected_content: 'kernel',
  kernel_error: 'kernel',
  harness_error: 'testbed',
};

/** Categories of protocol conformance tests */
export type TestCategory =
  | 'tier1_basic'
  | 'tier2_interactive'
  | 'tier3_rich_output'
  | 'tier4_advanced';

/** Human-readable tier descriptions */
export const TIER_DESCRIPTIONS: Record<TestCategory, string> = {
  tier1_basic: 'Basic Protocol',
  tier2_interactive: 'Interactive Features',
  tier3_rich_output: 'Rich Output',
  tier4_advanced: 'Advanced Features',
};

/** Tier numbers for sorting */
export const TIER_NUMBERS: Record<TestCategory, number> = {
  tier1_basic: 1,
  tier2_interactive: 2,
  tier3_rich_output: 3,
  tier4_advanced: 4,
};

/** Status of a test result */
export type TestStatus = 'pass' | 'fail' | 'unsupported' | 'timeout' | 'partial_pass';

/** Result of a single test execution (tagged union) */
export type TestResult =
  | { status: 'pass' }
  | { status: 'fail'; reason: string; kind?: FailureKind }
  | { status: 'unsupported' }
  | { status: 'timeout' }
  | { status: 'partial_pass'; score: number; notes: string };

/** Get the status emoji for a test result */
export function getStatusEmoji(status: TestStatus): string {
  switch (status) {
    case 'pass':
      return '✅';
    case 'fail':
      return '❌';
    case 'unsupported':
      return '⏭️';
    case 'timeout':
      return '⏱️';
    case 'partial_pass':
      return '⚠️';
  }
}

/** Record of a single test execution */
export interface TestRecord {
  /** Name of the test */
  name: string;
  /** Category/tier of the test */
  category: TestCategory;
  /** Human-readable description */
  description: string;
  /** The primary protocol message type being tested */
  message_type: string;
  /** Result of the test */
  result: TestResult;
  /** How long the test took in milliseconds */
  duration: number;
}

/** Report for a single kernel's conformance test run */
export interface KernelReport {
  /** Name of the kernel (e.g., "python3", "rust") */
  kernel_name: string;
  /** Language the kernel executes */
  language: string;
  /** Implementation name (e.g., "ipykernel") */
  implementation: string;
  /** Protocol version reported by kernel */
  protocol_version: string;
  /** Individual test results */
  results: TestRecord[];
  /** When the test run started (ISO 8601) */
  timestamp: string;
  /** Total duration of test run in milliseconds */
  total_duration: number;
}

/** Matrix of conformance results across multiple kernels */
export interface ConformanceMatrix {
  /** Reports from each kernel tested */
  reports: KernelReport[];
  /** When the matrix was generated (ISO 8601) */
  generated_at: string;
  /** Git commit SHA (added by CI) */
  commit_sha?: string;
}

/** Helper functions for working with reports */

export function getPassedCount(report: KernelReport): number {
  return report.results.filter((r) => r.result.status === 'pass' || r.result.status === 'partial_pass').length;
}

export function getTotalCount(report: KernelReport): number {
  return report.results.length;
}

export function getScore(report: KernelReport): number {
  const total = getTotalCount(report);
  if (total === 0) return 0;
  return getPassedCount(report) / total;
}

export function getTierResults(report: KernelReport, tier: TestCategory): TestRecord[] {
  return report.results.filter((r) => r.category === tier);
}

export function getTierScore(report: KernelReport, tier: TestCategory): [number, number] {
  const results = getTierResults(report, tier);
  const passed = results.filter((r) => r.result.status === 'pass' || r.result.status === 'partial_pass').length;
  return [passed, results.length];
}

export function getAllTestNames(matrix: ConformanceMatrix): string[] {
  const names = new Set<string>();
  for (const report of matrix.reports) {
    for (const test of report.results) {
      names.add(test.name);
    }
  }
  return Array.from(names).sort();
}
